<a href="https://colab.research.google.com/github/SintiyaMaharani34/zomato-delivery-performance-analysis-/blob/main/1_Data_Preparation_Cleaning_Zomato_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ==========================================
# PHASE 2 - DATA UNDERSTANDING
# ==========================================

In [ ]:
# Import Libraries

# Data Manipulation
import pandas as pd
import numpy as np

# Data Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Ignore Warning
import warnings
warnings.filterwarnings('ignore')

# Display Setting
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

# Visualization Style
plt.style.use('ggplot')
sns.set_theme(style='whitegrid')

print("Libraries imported successfully!")

Libraries imported successfully!


In [ ]:
# ==========================================
# Upload Dataset
# ==========================================

from google.colab import files

uploaded = files.upload()

Saving Zomato Dataset.csv to Zomato Dataset (5).csv


In [ ]:
# ==========================================
# Load Dataset
# ==========================================

import io

df = pd.read_csv(io.BytesIO(next(iter(uploaded.values()))))

print("Dataset loaded successfully!")

Dataset loaded successfully!


In [ ]:
# Display first 5 rows
df.head()

,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,Time_Order_picked,Weather_conditions,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken (min)
0,0xcdcd,DEHRES17DEL01,36.0,4.2,30.327968,78.046106,30.397968,78.116106,12-02-2022,21:55,22:10,Fog,Jam,2,Snack,motorcycle,3.0,No,Metropolitian,46
1,0xd987,KOCRES16DEL01,21.0,4.7,10.003064,76.307589,10.043064,76.347589,13-02-2022,14:55,15:05,Stormy,High,1,Meal,motorcycle,1.0,No,Metropolitian,23
2,0x2784,PUNERES13DEL03,23.0,4.7,18.562450,73.916619,18.652450,74.006619,04-03-2022,17:30,17:40,Sandstorms,Medium,1,Drinks,scooter,1.0,No,Metropolitian,21
3,0xc8b6,LUDHRES15DEL02,34.0,4.3,30.899584,75.809346,30.919584,75.829346,13-02-2022,09:20,09:30,Sandstorms,Low,0,Buffet,motorcycle,0.0,No,Metropolitian,20
4,0xdb64,KNPRES14DEL02,24.0,4.7,26.463504,80.372929,26.593504,80.502929,14-02-2022,19:50,20:05,Fog,Jam,1,Snack,scooter,1.0,No,Metropolitian,41


In [ ]:
# Display last 5 rows
df.tail()

,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,Time_Order_picked,Weather_conditions,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken (min)
45579,0x7c09,JAPRES04DEL01,30.0,4.8,26.902328,75.794257,26.912328,75.804257,24-03-2022,11:35,11:45,Windy,High,1,Meal,motorcycle,0.0,No,Metropolitian,32
45580,0xd641,AGRRES16DEL01,21.0,4.6,0.000000,0.000000,0.070000,0.070000,16-02-2022,19:55,20:10,Windy,Jam,0,Buffet,motorcycle,1.0,No,Metropolitian,36
45581,0x4f8d,CHENRES08DEL03,30.0,4.9,13.022394,80.242439,13.052394,80.272439,11-03-2022,23:50,24:05:00,Cloudy,Low,1,Drinks,scooter,0.0,No,Metropolitian,16
45582,0x5eee,COIMBRES11DEL01,20.0,4.7,11.001753,76.986241,11.041753,77.026241,07-03-2022,13:35,13:40,Cloudy,High,0,Snack,motorcycle,1.0,No,Metropolitian,26
45583,0x5fb2,RANCHIRES09DEL02,23.0,4.9,23.351058,85.325731,23.431058,85.405731,02-03-2022,17:10,17:15,Fog,Medium,2,Snack,scooter,1.0,No,Metropolitian,36


In [ ]:
# Dataset Shape
print(f"Number of Rows    : {df.shape[0]}")
print(f"Number of Columns : {df.shape[1]}")

Number of Rows    : 45584
Number of Columns : 20


In [ ]:
df.columns

Index(['ID', 'Delivery_person_ID', 'Delivery_person_Age',
       'Delivery_person_Ratings', 'Restaurant_latitude',
       'Restaurant_longitude', 'Delivery_location_latitude',
       'Delivery_location_longitude', 'Order_Date', 'Time_Orderd',
       'Time_Order_picked', 'Weather_conditions', 'Road_traffic_density',
       'Vehicle_condition', 'Type_of_order', 'Type_of_vehicle',
       'multiple_deliveries', 'Festival', 'City', 'Time_taken (min)'],
      dtype='object')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45584 entries, 0 to 45583
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID                           45584 non-null  object 
 1   Delivery_person_ID           45584 non-null  object 
 2   Delivery_person_Age          43730 non-null  float64
 3   Delivery_person_Ratings      43676 non-null  float64
 4   Restaurant_latitude          45584 non-null  float64
 5   Restaurant_longitude         45584 non-null  float64
 6   Delivery_location_latitude   45584 non-null  float64
 7   Delivery_location_longitude  45584 non-null  float64
 8   Order_Date                   45584 non-null  object 
 9   Time_Orderd                  43853 non-null  object 
 10  Time_Order_picked            45584 non-null  object 
 11  Weather_conditions           44968 non-null  object 
 12  Road_traffic_density         44983 non-null  object 
 13  Vehicle_conditio

In [ ]:
#Numerical
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Delivery_person_Age,43730.0,29.566911,5.815064,15.000000,25.000000,30.000000,35.000000,50.000000
Delivery_person_Ratings,43676.0,4.633774,0.334744,1.000000,4.500000,4.700000,4.900000,6.000000
Restaurant_latitude,45584.0,17.017948,8.185674,-30.905562,12.933284,18.551440,22.728163,30.914057
Restaurant_longitude,45584.0,70.229684,22.885575,-88.366217,73.170000,75.897963,78.044095,88.433452
Delivery_location_latitude,45584.0,17.465480,7.335562,0.010000,12.988453,18.633934,22.785049,31.054057
Delivery_location_longitude,45584.0,70.844161,21.120578,0.010000,73.280000,76.002574,78.107044,88.563452
Vehicle_condition,45584.0,1.023385,0.839055,0.000000,0.000000,1.000000,2.000000,3.000000
multiple_deliveries,44591.0,0.744635,0.572510,0.000000,0.000000,1.000000,1.000000,3.000000
Time_taken (min),45584.0,26.293963,9.384298,10.000000,19.000000,26.000000,32.000000,54.000000


In [ ]:
#Categorical
df.describe(include='object').T

,count,unique,top,freq
ID,45584,45584,0x2bed,1
Delivery_person_ID,45584,1320,PUNERES01DEL01,67
Order_Date,45584,44,15-03-2022,1192
Time_Orderd,43853,176,21:55,461
Time_Order_picked,45584,193,21:30,495
Weather_conditions,44968,6,Fog,7653
Road_traffic_density,44983,4,Low,15476
Type_of_order,45584,4,Snack,11530
Type_of_vehicle,45584,4,motorcycle,26429
Festival,45356,2,No,44460


#DATA PREPARATION

1. Check Missing Value

In [ ]:
missing = pd.DataFrame({
    'Missing Values': df.isnull().sum(),
    'Percentage (%)': (df.isnull().sum() / len(df)) * 100
})

missing[missing['Missing Values'] > 0].sort_values(
    by='Percentage (%)',
    ascending=False
)

,Missing Values,Percentage (%)
Delivery_person_Ratings,1908,4.185679
Delivery_person_Age,1854,4.067217
Time_Orderd,1731,3.797385
City,1200,2.632503
multiple_deliveries,993,2.178396
Weather_conditions,616,1.351351
Road_traffic_density,601,1.318445
Festival,228,0.500176


2. Check Duplicate

In [ ]:
duplicate = df.duplicated().sum()

print(f"Duplicate Rows : {duplicate}")

Duplicate Rows : 0


3. Check Unique Value

In [ ]:
unique = pd.DataFrame({
    'Unique Values': df.nunique()
})

unique

,Unique Values
ID,45584
Delivery_person_ID,1320
Delivery_person_Age,22
Delivery_person_Ratings,28
Restaurant_latitude,657
Restaurant_longitude,518
Delivery_location_latitude,4373
Delivery_location_longitude,4373
Order_Date,44
Time_Orderd,176


In [ ]:
df.sample(10, random_state=42)

,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,Time_Order_picked,Weather_conditions,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken (min)
27811,0xcfcd,ALHRES08DEL02,32.0,4.1,25.450329,81.834279,25.500329,81.884279,13-02-2022,17:15,17:30,Sandstorms,Medium,0,Snack,motorcycle,1.0,No,Metropolitian,33
43475,0x4f03,MYSRES17DEL01,32.0,4.9,12.326356,76.619103,12.336356,76.629103,03-03-2022,08:35,08:45,Windy,Low,1,Snack,motorcycle,1.0,No,Metropolitian,28
11902,0xc48b,DEHRES01DEL01,23.0,4.5,0.000000,0.000000,0.070000,0.070000,14-02-2022,21:50,22:05,Sunny,Jam,2,Snack,scooter,1.0,No,Metropolitian,22
36731,0x277c,JAPRES20DEL03,22.0,4.5,26.956431,75.776649,26.986431,75.806649,13-03-2022,22:45,22:50,Stormy,Low,0,Snack,motorcycle,1.0,No,Metropolitian,25
18188,0x5097,MYSRES04DEL02,29.0,4.9,12.308500,76.665808,12.438500,76.795808,14-03-2022,21:20,21:35,Cloudy,Jam,0,Drinks,motorcycle,1.0,No,Metropolitian,35
28498,0xdb9a,LUDHRES08DEL02,36.0,4.7,30.899992,75.831338,31.029992,75.961338,18-02-2022,23:25,23:30,Fog,Low,2,Drinks,motorcycle,0.0,No,Metropolitian,27
4740,0x22e6,MYSRES01DEL01,20.0,3.6,12.311072,76.654878,12.321072,76.664878,03-03-2022,11:15,11:20,Sunny,High,0,Meal,motorcycle,1.0,No,Metropolitian,48
6868,0x378f,BANGRES14DEL01,34.0,4.8,12.949934,77.699386,12.989934,77.739386,21-03-2022,16:35,16:50,Fog,Medium,0,Snack,motorcycle,1.0,No,Metropolitian,27
43338,0xa647,JAPRES05DEL01,27.0,4.7,26.905287,75.794592,27.015287,75.904592,27-03-2022,20:50,20:55,Stormy,Jam,1,Meal,scooter,0.0,No,Urban,24
35602,0x864c,MUMRES14DEL01,35.0,4.7,19.181300,72.836191,19.251300,72.906191,02-04-2022,18:15,18:25,Sunny,Medium,2,Drinks,scooter,0.0,No,Metropolitian,18


DATASET SUMMARY

In [ ]:
summary = pd.DataFrame({
    'Data Type': df.dtypes,
    'Missing Values': df.isnull().sum(),
    'Unique Values': df.nunique()
})

summary

,Data Type,Missing Values,Unique Values
ID,object,0,45584
Delivery_person_ID,object,0,1320
Delivery_person_Age,float64,1854,22
Delivery_person_Ratings,float64,1908,28
Restaurant_latitude,float64,0,657
Restaurant_longitude,float64,0,518
Delivery_location_latitude,float64,0,4373
Delivery_location_longitude,float64,0,4373
Order_Date,object,0,44
Time_Orderd,object,1731,176


#Initial Observation
Berdasarkan hasil data understanding, ditemukan beberapa permasalahan kualitas data yang perlu ditangani sebelum analisis, antara lain:

Missing values pada beberapa kolom penting seperti Delivery_person_Age, Delivery_person_Ratings, dan Time_Ordered.
Beberapa kolom bertipe object perlu dikonversi menjadi datetime agar dapat dianalisis berdasarkan waktu.
Dataset memiliki kombinasi variabel numerik dan kategorikal yang cukup lengkap untuk melakukan analisis hubungan antara karakteristik pengiriman dan delivery time sebagai target utama.

#DATA CLEANING

In [ ]:
# Check unique values for object columns

for col in df.select_dtypes(include='object').columns:
    print(f"\n===== {col} =====")
    print(df[col].unique()[:15])


===== ID =====
['0xcdcd' '0xd987' '0x2784' '0xc8b6' '0xdb64' '0x3af3' '0x3aab' '0x689b'
 '0x6f67' '0xc9cf' '0x36b8' '0x5795' '0x6c6b' '0xb816' '0x539b']

===== Delivery_person_ID =====
['DEHRES17DEL01' 'KOCRES16DEL01' 'PUNERES13DEL03' 'LUDHRES15DEL02'
 'KNPRES14DEL02' 'MUMRES15DEL03' 'MYSRES01DEL01' 'PUNERES20DEL01'
 'HYDRES14DEL01' 'KOLRES15DEL03' 'PUNERES19DEL02' 'RANCHIRES06DEL02'
 'COIMBRES13DEL01' 'CHENRES19DEL02' 'MUMRES02DEL01']

===== Order_Date =====
['12-02-2022' '13-02-2022' '04-03-2022' '14-02-2022' '02-04-2022'
 '01-03-2022' '16-03-2022' '20-03-2022' '15-02-2022' '10-03-2022'
 '11-03-2022' '27-03-2022' '26-03-2022' '05-04-2022' '12-03-2022']

===== Time_Orderd =====
['21:55' '14:55' '17:30' '09:20' '19:50' '20:25' '20:30' '20:40' '21:15'
 '20:20' '22:30' '08:15' '19:30' '12:25' '18:35']

===== Time_Order_picked =====
['22:10' '15:05' '17:40' '09:30' '20:05' '20:35' '15:10' '20:40' '20:50'
 '21:30' '20:25' '22:45' '08:30' '19:45' '12:30']

===== Weather_conditions =====
['

In [ ]:
# Remove leading and trailing spaces

for col in df.select_dtypes(include='object').columns:
    df[col] = df[col].str.strip()

Handling Missing Value

In [ ]:
missing = pd.DataFrame({
    'Missing Values': df.isnull().sum(),
    'Percentage (%)': round(df.isnull().mean()*100,2)
})

missing[missing['Missing Values']>0].sort_values(
    by='Missing Values',
    ascending=False
)

,Missing Values,Percentage (%)
Delivery_person_Ratings,1908,4.19
Delivery_person_Age,1854,4.07
Time_Orderd,1731,3.80
City,1200,2.63
multiple_deliveries,993,2.18
Weather_conditions,616,1.35
Road_traffic_density,601,1.32
Festival,228,0.50


In [ ]:
#Numerical (menggunakan Median)

num_cols = [
    'Delivery_person_Age',
    'Delivery_person_Ratings',
    'multiple_deliveries'
]

for col in num_cols:
    df[col].fillna(df[col].median(), inplace=True)

In [ ]:
#Categorical (Menggunakan Modus)

cat_cols = [
    'Weather_conditions',
    'Road_traffic_density',
    'Festival',
    'City'
]

for col in cat_cols:
    df[col].fillna(df[col].mode()[0], inplace=True)

#Reason

Kolom Time_Orderd merupakan komponen utama dalam analisis waktu pemesanan ( Order_Datetime) dan durasi persiapan pesanan (Preparation_Time). Baris dengan nilai yang hilang dihapus agar metrik berbasis waktu yang dihasilkan tetap akurat dan dapat diandalkan untuk mendukung pengambilan keputusan. Keputusan ini diambil untuk menghindari bias serta menjaga konsistensi analisis pada tahap selanjutnya.

In [ ]:
df.isnull().sum().sort_values(ascending=False)

,0
Time_Orderd,1731
ID,0
Delivery_person_Age,0
Delivery_person_ID,0
Delivery_person_Ratings,0
Restaurant_latitude,0
Delivery_location_latitude,0
Restaurant_longitude,0
Delivery_location_longitude,0
Order_Date,0


In [ ]:
print("Before :", df.shape)

df = df.dropna(subset=['Time_Orderd']).reset_index(drop=True)

print("After  :", df.shape)

print(df['Time_Orderd'].isnull().sum())

Before : (45584, 20)
After  : (43853, 20)
0


In [ ]:
df.isnull().sum()

,0
ID,0
Delivery_person_ID,0
Delivery_person_Age,0
Delivery_person_Ratings,0
Restaurant_latitude,0
Restaurant_longitude,0
Delivery_location_latitude,0
Delivery_location_longitude,0
Order_Date,0
Time_Orderd,0


In [ ]:
#Convert Data Type

df['Order_Date'] = pd.to_datetime(
    df['Order_Date'],
    format='%d-%m-%Y'
)

In [ ]:
df[['Time_Orderd','Time_Order_picked']].head()

,Time_Orderd,Time_Order_picked
0,21:55,22:10
1,14:55,15:05
2,17:30,17:40
3,09:20,09:30
4,19:50,20:05


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43853 entries, 0 to 43852
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   ID                           43853 non-null  object        
 1   Delivery_person_ID           43853 non-null  object        
 2   Delivery_person_Age          43853 non-null  float64       
 3   Delivery_person_Ratings      43853 non-null  float64       
 4   Restaurant_latitude          43853 non-null  float64       
 5   Restaurant_longitude         43853 non-null  float64       
 6   Delivery_location_latitude   43853 non-null  float64       
 7   Delivery_location_longitude  43853 non-null  float64       
 8   Order_Date                   43853 non-null  datetime64[ns]
 9   Time_Orderd                  43853 non-null  object        
 10  Time_Order_picked            43853 non-null  object        
 11  Weather_conditions           43853 non-nu

In [ ]:
#convert Numerical Columns

numeric_cols = [
    'Delivery_person_Age',
    'Delivery_person_Ratings',
    'multiple_deliveries'
]

for col in numeric_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43853 entries, 0 to 43852
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   ID                           43853 non-null  object        
 1   Delivery_person_ID           43853 non-null  object        
 2   Delivery_person_Age          43853 non-null  float64       
 3   Delivery_person_Ratings      43853 non-null  float64       
 4   Restaurant_latitude          43853 non-null  float64       
 5   Restaurant_longitude         43853 non-null  float64       
 6   Delivery_location_latitude   43853 non-null  float64       
 7   Delivery_location_longitude  43853 non-null  float64       
 8   Order_Date                   43853 non-null  datetime64[ns]
 9   Time_Orderd                  43853 non-null  object        
 10  Time_Order_picked            43853 non-null  object        
 11  Weather_conditions           43853 non-nu

| Kolom                   | Missing   | Treatment     | Alasan                                                                                          |
| ----------------------- | --------- | ------------- | ----------------------------------------------------------------------------------------------- |
| Delivery_person_Age     | 4.07%     | Median        | Variabel numerik, masih dapat diestimasi                                                        |
| Delivery_person_Ratings | 4.19%     | Median        | Variabel numerik, distribusi relatif stabil                                                     |
| Weather_conditions      | 1.35%     | Mode          | Variabel kategorikal                                                                            |
| Road_traffic_density    | 1.32%     | Mode          | Variabel kategorikal                                                                            |
| Festival                | 0.50%     | Mode          | Variabel kategorikal                                                                            |
| City                    | 2.63%     | Mode          | Variabel kategorikal                                                                            |
| **Time_Orderd**         | **3.80%** | **Drop Rows** | **Tidak dapat digunakan dalam proses EDA, sehingga di drop** |


In [ ]:
print(df['Time_Orderd'].isna().sum())

print(df['Time_Orderd'].eq('').sum())

print(df['Time_Orderd'].astype(str).eq('NaN').sum())

print(df['Time_Orderd'].astype(str).eq('nan').sum())

print(df['Time_Orderd'].astype(str).eq('None').sum())

print(df['Time_Orderd'].astype(str).eq(' ').sum())

0
0
0
0
0
0


In [ ]:
print(df.loc[113, 'Time_Orderd'])
print(type(df.loc[113, 'Time_Orderd']))

1
<class 'str'>


# ==========================================
# PHASE 3 - FEATURE ENGINEERING
# ==========================================

📌 Business Purpose

Create meaningful features that help identify operational factors affecting delivery performance and support further exploratory data analysis.

In [ ]:
# ==========================================
# Check Time Format
# ==========================================

print("Unique Time_Orderd (sample):")
print(df['Time_Orderd'].drop_duplicates().sample(20, random_state=42).tolist())

print("\nUnique Time_Order_picked (sample):")
print(df['Time_Order_picked'].drop_duplicates().sample(20, random_state=42).tolist())

Unique Time_Orderd (sample):
['22:35', '22:25', '17:40', '19:45', '10:50', '23:20', '11:10', '13:55', '10:30', '14:40', '14:15', '0.541666667', '22:50', '18:40', '20:35', '15:35', '22:55', '18:45', '20:20', '0.666666667']

Unique Time_Order_picked (sample):
['20:20', '15:35', '20:45', '13:45', '15:50', '0.416666667', '18:50', '23:45', '23:30', '21:30', '23:15', '09:50', '12:50', '11:50', '11:35', '17:30', '08:45', '10:20', '18:10', '21:40']


In [ ]:
# ==========================================
# Time Standardization Function
# ==========================================

from datetime import datetime, timedelta
import pandas as pd
import numpy as np

def standardize_time(value):

    # Missing value
    if pd.isna(value):
        return np.nan

    value = str(value).strip()

    # ====================================
    # Case 1 : Excel Serial Time
    # ====================================
    try:
        if ":" not in value:
            excel_time = float(value)
            total_seconds = round(excel_time * 86400)
            hour = total_seconds // 3600
            minute = (total_seconds % 3600) // 60
            hour = hour % 24
            return f"{hour:02d}:{minute:02d}"
    except:
        pass

    # ====================================
    # Case 2 : HH:MM / HH:MM:SS
    # ====================================
    if ":" in value:
        parts = value.split(":")
        hour = int(parts[0])
        minute = int(parts[1])
        if hour == 24:
            hour = 0
        return f"{hour:02d}:{minute:02d}"
    return np.nan

In [ ]:
# ==========================================
# Standardize Time (Overwrite Original)
# ==========================================

df['Time_Orderd'] = df['Time_Orderd'].apply(standardize_time)
df['Time_Order_picked'] = df['Time_Order_picked'].apply(standardize_time)

In [ ]:
# Cek masih ada Excel serial time atau tidak
mask = df['Time_Orderd'].astype(str).str.contains(r'\.', regex=True)

print("Remaining Excel Serial Time:", mask.sum())

df.loc[mask, ['Time_Orderd']].head()

Remaining Excel Serial Time: 0


,Time_Orderd


In [ ]:
print(df['Time_Orderd'].isna().sum())
print(df['Time_Order_picked'].isna().sum())

0
0


In [ ]:
# Identify records with 24:xx

pickup_next_day = df['Time_Order_picked'].astype(str).str.startswith('24:')

In [ ]:
print(df['Time_Orderd'].head())

print(df['Time_Order_picked'].head())

0    21:55
1    14:55
2    17:30
3    09:20
4    19:50
Name: Time_Orderd, dtype: object
0    22:10
1    15:05
2    17:40
3    09:30
4    20:05
Name: Time_Order_picked, dtype: object


In [ ]:
# ==========================================
# Create Order Datetime
# ==========================================

df['Order_Datetime'] = pd.to_datetime(
    df['Order_Date'].dt.strftime('%Y-%m-%d') + ' ' + df['Time_Orderd'],
    format='%Y-%m-%d %H:%M'
)

In [ ]:
df[['Order_Date','Time_Orderd','Order_Datetime']].head()

,Order_Date,Time_Orderd,Order_Datetime
0,2022-02-12,21:55,2022-02-12 21:55:00
1,2022-02-13,14:55,2022-02-13 14:55:00
2,2022-03-04,17:30,2022-03-04 17:30:00
3,2022-02-13,09:20,2022-02-13 09:20:00
4,2022-02-14,19:50,2022-02-14 19:50:00


In [ ]:
# ==========================================
# Create Pickup Datetime
# ==========================================

df['Pickup_Datetime'] = pd.to_datetime(
    df['Order_Date'].dt.strftime('%Y-%m-%d') + ' ' + df['Time_Order_picked'],
    format='%Y-%m-%d %H:%M'
)

In [ ]:
df[['Time_Order_picked','Pickup_Datetime']].head()

,Time_Order_picked,Pickup_Datetime
0,22:10,2022-02-12 22:10:00
1,15:05,2022-02-13 15:05:00
2,17:40,2022-03-04 17:40:00
3,09:30,2022-02-13 09:30:00
4,20:05,2022-02-14 20:05:00


In [ ]:
# ==========================================
# Handle Pickup After Midnight
# ==========================================

time_diff = df['Order_Datetime'] - df['Pickup_Datetime']

mask = (
    (df['Pickup_Datetime'] < df['Order_Datetime']) &
    (time_diff > pd.Timedelta(hours=12))
)

df.loc[mask, 'Pickup_Datetime'] += pd.Timedelta(days=1)

In [ ]:
# ==========================================
# Create Preparation Time
# ==========================================

df['Preparation_Time'] = (

    df['Pickup_Datetime']
    -
    df['Order_Datetime']

).dt.total_seconds() / 60

In [ ]:
df['Preparation_Time'].describe()

,Preparation_Time
count,43853.000000
mean,9.989168
std,4.087516
min,5.000000
25%,5.000000
50%,10.000000
75%,15.000000
max,15.000000


In [ ]:
df['Preparation_Time'].value_counts().sort_index()

,count
Preparation_Time,
5.0,14701
10.0,14546
15.0,14606


#==========================================
#Feature 2 : Delivery Distance

#==========================================



In [ ]:
# ==========================================
# Coordinate Summary
# ==========================================

coordinate_cols = [
    'Restaurant_latitude',
    'Restaurant_longitude',
    'Delivery_location_latitude',
    'Delivery_location_longitude'
]

df[coordinate_cols].describe().T

,count,mean,std,min,25%,50%,75%,max
Restaurant_latitude,43853.0,17.242245,7.699200,-30.902872,12.933298,18.554382,22.732225,30.914057
Restaurant_longitude,43853.0,70.763328,21.138033,0.000000,73.170283,75.898497,78.045359,88.433452
Delivery_location_latitude,43853.0,17.462771,7.338998,0.010000,12.986229,18.633934,22.785049,31.054057
Delivery_location_longitude,43853.0,70.826921,21.138202,0.010000,73.279083,75.999490,78.103162,88.563452


In [ ]:
df[coordinate_cols].isnull().sum()

,0
Restaurant_latitude,0
Restaurant_longitude,0
Delivery_location_latitude,0
Delivery_location_longitude,0


In [ ]:
negative_coord = df[
    (df['Restaurant_latitude'] < 0) |
    (df['Restaurant_longitude'] < 0) |
    (df['Delivery_location_latitude'] < 0) |
    (df['Delivery_location_longitude'] < 0)
]

print(f"Negative Coordinate Records : {len(negative_coord)}")
print(f"Percentage : {round(len(negative_coord)/len(df)*100,2)}%")

Negative Coordinate Records : 156
Percentage : 0.36%


In [ ]:
# ==========================================
# Haversine Formula
# ==========================================

from math import radians, sin, cos, sqrt, atan2

def haversine(lat1, lon1, lat2, lon2):

    R = 6371

    lat1 = radians(lat1)
    lon1 = radians(lon1)

    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlat = lat2 - lat1
    dlon = lon2 - lon1

    a = (
        sin(dlat/2)**2
        +
        cos(lat1) *
        cos(lat2) *
        sin(dlon/2)**2
    )

    c = 2 * atan2(
        sqrt(a),
        sqrt(1-a)
    )

    return R * c

In [ ]:
# ==========================================
# Delivery Distance (KM)
# ==========================================

df['Delivery_Distance_km'] = df.apply(

    lambda row: haversine(

        row['Restaurant_latitude'],
        row['Restaurant_longitude'],

        row['Delivery_location_latitude'],
        row['Delivery_location_longitude']

    ),

    axis=1
)

In [ ]:
df['Delivery_Distance_km'].describe()

,Delivery_Distance_km
count,43853.000000
mean,27.173498
std,303.107387
min,1.465067
25%,4.663432
50%,9.220419
75%,13.682165
max,6884.726399


In [ ]:
df.nlargest(
    20,
    'Delivery_Distance_km'
)[[
    'Restaurant_latitude',
    'Restaurant_longitude',
    'Delivery_location_latitude',
    'Delivery_location_longitude',
    'Delivery_Distance_km'
]]

,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Delivery_Distance_km
3900,-30.902872,75.826808,31.012872,75.936808,6884.726399
2779,-30.895817,75.813112,31.005817,75.923112,6883.157441
2523,-30.902872,75.826808,30.972872,75.896808,6880.272782
24934,-30.895817,75.813112,30.965817,75.883112,6878.703822
22324,-30.899584,75.809346,30.959584,75.869346,6878.428563
18195,-30.885814,75.786976,30.965814,75.866976,6877.592421
25689,-30.899584,75.809346,30.949584,75.859346,6877.315724
23891,-30.873988,75.842739,30.963988,75.932739,6876.075767
9473,-30.885915,75.788259,30.945915,75.848259,6875.388718
7913,-30.899584,75.809346,30.929584,75.839346,6875.090529


In [ ]:
df.sort_values(
    by='Delivery_Distance_km',
    ascending=False
)[[
    'Restaurant_latitude',
    'Restaurant_longitude',
    'Delivery_location_latitude',
    'Delivery_location_longitude',
    'Delivery_Distance_km'
]].head(20)

,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Delivery_Distance_km
3900,-30.902872,75.826808,31.012872,75.936808,6884.726399
2779,-30.895817,75.813112,31.005817,75.923112,6883.157441
2523,-30.902872,75.826808,30.972872,75.896808,6880.272782
24934,-30.895817,75.813112,30.965817,75.883112,6878.703822
22324,-30.899584,75.809346,30.959584,75.869346,6878.428563
18195,-30.885814,75.786976,30.965814,75.866976,6877.592421
25689,-30.899584,75.809346,30.949584,75.859346,6877.315724
23891,-30.873988,75.842739,30.963988,75.932739,6876.075767
9473,-30.885915,75.788259,30.945915,75.848259,6875.388718
7913,-30.899584,75.809346,30.929584,75.839346,6875.090529


In [ ]:
print(df['Restaurant_latitude'].nunique())

print(df[df['Restaurant_latitude'] < 0]['Restaurant_latitude'].unique())

491
[-27.163303 -27.165108 -15.546594 -23.230791 -22.539129 -15.157944
  -9.959778 -15.51315  -15.576683 -30.902872  -9.988483 -23.211529
 -30.895817 -26.474986 -15.5696   -26.472001 -25.449872 -22.553227
  -9.966783 -19.878028  -9.970717 -22.526461 -19.876994 -25.449659
 -25.450317 -26.49095  -30.319528 -10.028047 -19.876106 -26.463504
 -30.899584 -23.234631 -23.234249 -15.496162 -22.538999 -30.885915
 -22.53796  -23.184992 -10.035573 -23.235123 -25.453436 -23.214459
 -26.483042 -19.879631 -19.874449 -10.020683 -19.875337 -19.875522
 -19.875016 -22.569358 -27.201725 -22.538731 -27.160832 -30.885814
 -25.451517 -22.547186 -26.474133 -30.327968 -26.481547 -22.515082
 -26.473698 -30.873988 -22.514119 -19.876428  -9.960846 -15.303897
 -27.16185  -30.328174 -22.5491   -26.471529 -30.893244 -19.888716
 -22.551084 -15.561295  -9.979186 -30.346994 -25.454697 -30.372202
 -19.880256 -10.027014 -30.361281 -10.003064 -22.552996  -9.982834
 -19.866969 -22.577821 -22.569367 -23.232357 -19.875908 -2

In [ ]:
negative_coord = df[df['Restaurant_latitude'] < 0]

In [ ]:
negative_coord[['Delivery_Distance_km']].describe()

,Delivery_Distance_km
count,156.000000
mean,4912.290324
std,1371.032845
min,2216.065794
25%,4421.166849
50%,5021.474291
75%,5895.928471
max,6884.726399


In [ ]:
negative_coord[['Restaurant_latitude',
                'Delivery_location_latitude',
                'Delivery_Distance_km']].head(10)

,Restaurant_latitude,Delivery_location_latitude,Delivery_Distance_km
81,-27.163303,27.233303,6048.631240
267,-27.165108,27.225108,6047.919478
1039,-15.546594,15.606594,3464.082720
1701,-23.230791,23.250791,5168.516553
1889,-22.539129,22.549129,5013.585658
2011,-15.157944,15.197944,3375.423601
2048,-9.959778,9.969778,2216.065794
2108,-15.513150,15.563150,3455.531261
2199,-15.576683,15.646683,3471.888400
2523,-30.902872,30.972872,6880.272782


In [ ]:
print(df['Delivery_location_latitude'].min())
print(df['Delivery_location_latitude'].max())

0.01
31.054057


### Coordinate Validation

During feature engineering, 156 records produced unrealistic delivery distances ranging from approximately 2,200 km to 6,900 km.

Further investigation showed that these records contained inconsistent geographic coordinates, where restaurant latitude was negative while delivery latitude remained positive, resulting in impossible delivery routes for a food delivery service.

Since the correct coordinates cannot be inferred, these records were treated as data quality issues and removed from further analysis.

In [ ]:
# ==========================================
# Remove Invalid Geographic Coordinates
# ==========================================

df = df[df['Restaurant_latitude'] >= 0].copy()

print(f"Remaining Records : {len(df):,}")

Remaining Records : 43,697


In [ ]:
# Hapus kolom lama
df.drop(columns=['Delivery_Distance_km'], inplace=True, errors='ignore')

# Hitung ulang
df['Delivery_Distance_km'] = df.apply(
    lambda row: haversine(
        row['Restaurant_latitude'],
        row['Restaurant_longitude'],
        row['Delivery_location_latitude'],
        row['Delivery_location_longitude']
    ),
    axis=1
)

In [ ]:
df['Delivery_Distance_km'].describe()

,Delivery_Distance_km
count,43697.000000
mean,9.733440
std,5.604191
min,1.465067
25%,4.663412
50%,9.220209
75%,13.681416
max,20.969489


In [ ]:
# ==========================================
# Order Hour
# ==========================================

df['Order_Hour'] = df['Order_Datetime'].dt.hour

In [ ]:
# ==========================================
# Day_of_Week
# ==========================================

df['Day_of_Week'] = df['Order_Datetime'].dt.day_name()

In [ ]:
# ==========================================
# Weekend Flag
# ==========================================

df['Is_Weekend'] = np.where(
    df['Day_of_Week'].isin(['Saturday', 'Sunday']),
    'Weekend',
    'Weekday'
)

In [ ]:
# ==========================================
# Time of Day
# ==========================================

def time_of_day(hour):

    if 5 <= hour < 12:
        return 'Morning'

    elif 12 <= hour < 17:
        return 'Afternoon'

    elif 17 <= hour < 21:
        return 'Evening'

    else:
        return 'Night'

In [ ]:
df['Time_of_Day'] = df['Order_Hour'].apply(time_of_day)

In [ ]:
new_features = [

    'Order_Datetime',

    'Pickup_Datetime',

    'Preparation_Time',

    'Delivery_Distance_km',

    'Order_Hour',

    'Day_of_Week',

    'Is_Weekend',

    'Time_of_Day',

]

df[new_features].head()

,Order_Datetime,Pickup_Datetime,Preparation_Time,Delivery_Distance_km,Order_Hour,Day_of_Week,Is_Weekend,Time_of_Day
0,2022-02-12 21:55:00,2022-02-12 22:10:00,15.0,10.280582,21.0,Saturday,Weekend,Night
1,2022-02-13 14:55:00,2022-02-13 15:05:00,10.0,6.242319,14.0,Sunday,Weekend,Afternoon
2,2022-03-04 17:30:00,2022-03-04 17:40:00,10.0,13.787860,17.0,Friday,Weekday,Evening
3,2022-02-13 09:20:00,2022-02-13 09:30:00,10.0,2.930258,9.0,Sunday,Weekend,Morning
4,2022-02-14 19:50:00,2022-02-14 20:05:00,15.0,19.396618,19.0,Monday,Weekday,Evening


| Feature              | Business Purpose                                     |
| -------------------- | ---------------------------------------------------- |
| Preparation_Time     | Mengukur efisiensi restoran dalam menyiapkan pesanan |
| Delivery_Distance_km | Mengetahui pengaruh jarak terhadap waktu pengiriman  |             |
| Order_Hour           | Mengidentifikasi jam sibuk                           |
| Day_of_Week           | Membandingkan performa berdasarkan hari              |
| Is_Weekend              | Analisis weekday vs weekend                          |
| Time_of_Day          | Analisis pagi, siang, sore, malam                    |      |


# ==========================================
# PHASE 4 - FINAL VALIDATION
# ==========================================

In [ ]:
print("Dataset Shape :", df.shape)

Dataset Shape : (43697, 28)


In [ ]:
df.isnull().sum().sort_values(ascending=False).head(10)

,0
ID,0
Delivery_person_ID,0
Delivery_person_Age,0
Delivery_person_Ratings,0
Restaurant_latitude,0
Restaurant_longitude,0
Delivery_location_latitude,0
Delivery_location_longitude,0
Order_Date,0
Time_Orderd,0


In [ ]:
print("Duplicate Rows :", df.duplicated().sum())

Duplicate Rows : 0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43697 entries, 0 to 43852
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   ID                           43697 non-null  object        
 1   Delivery_person_ID           43697 non-null  object        
 2   Delivery_person_Age          43697 non-null  float64       
 3   Delivery_person_Ratings      43697 non-null  float64       
 4   Restaurant_latitude          43697 non-null  float64       
 5   Restaurant_longitude         43697 non-null  float64       
 6   Delivery_location_latitude   43697 non-null  float64       
 7   Delivery_location_longitude  43697 non-null  float64       
 8   Order_Date                   43697 non-null  datetime64[ns]
 9   Time_Orderd                  43697 non-null  object        
 10  Time_Order_picked            43697 non-null  object        
 11  Weather_conditions           43697 non-null  o

In [ ]:
print("="*50)
print("FINAL DATA VALIDATION")
print("="*50)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")

print("\nMissing Values")
print(df.isnull().sum().sum())

print("\nDuplicate Rows")
print(df.duplicated().sum())

FINAL DATA VALIDATION
Rows    : 43697
Columns : 28

Missing Values
0

Duplicate Rows
0


In [ ]:
# ==========================================
# Save Clean Dataset
# ==========================================

df.to_csv(
    "zomato_delivery_clean.csv",
    index=False
)

print("Dataset successfully saved!")

Dataset successfully saved!


In [ ]:
from google.colab import files

files.download("zomato_delivery_clean.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>